pageserver/src/pgdatadir_mapping.rs

```diff
@@ -1331,56 +1363,58 @@ impl Timeline {
 
     /// Get cached size of relation if it not updated after specified LSN
     pub fn get_cached_rel_size(&self, tag: &RelTag, lsn: Lsn) -> Option<BlockNumber> {
-        let rel_size_cache = self.rel_size_cache.read().unwrap();
-        if let Some((cached_lsn, nblocks)) = rel_size_cache.map.get(tag) {
+        let rel_size_cache = self.rel_size_primary_cache.read().unwrap();
+        if let Some((cached_lsn, nblocks)) = rel_size_cache.get(tag) {
             if lsn >= *cached_lsn {
                 RELSIZE_CACHE_HITS.inc();
                 return Some(*nblocks);
             }
             RELSIZE_CACHE_MISSES_OLD.inc();
         }
+        let mut rel_size_cache = self.rel_size_replica_cache.lock().unwrap();
+        if let Some(nblock) = rel_size_cache.get(&(lsn, *tag)) {
+            RELSIZE_CACHE_HITS.inc();
```

Comment:
We need a way of telling if this new cache is effective.
I propose:
* don't read from the replica cache if we are serving the primary compute
* separate hit and miss metrics for the caches

Reply:
> don't read from the replica cache if we are serving the primary compute

I am not sure here. Right now we may use `request_lsn != inf` even we are serving primary compute: when we extract page from parent branch.

Reply:
> I am not sure here. Right now we may use request_lsn != inf even we are serving primary compute: when we extract page from parent branch.

Thinking about this more. How so? `request_lsn` does not change when the read path descend into the ancestor and the comput does not know about branches.

Reply:
Sorry, please ignore my reply. Branching has actually no relation to request_lsn.
But still if static replica is created at the last_lsn and it access relation size before primary, then it makes sense to used value cached by replica.